### PR TITLE
Allow sizes of block and index caches to be updated.

### PR DESCRIPTION
**Pull request by martinmr**
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1551)

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -1981,3 +1981,45 @@ func (db *DB) MaxVersion() (uint64, error) {
...
 }
+
+type CacheType int
+
+const (
+	BlockCache CacheType = iota
+	IndexCache
+)
+
+func (db *DB) CacheMaxCost(cache CacheType) (int64, error) {
```
Comment (jarifibrahim) [inline db.go:1992]: Add a comment please since this is a public API.

File: db_test.go [hunk 2]
```diff
@@ -2184,3 +2184,32 @@ func TestWriteInemory(t *testing.T) {
...
+	require.Equal(t, int64(1<<20), cost)
+	cost, err = db.CacheMaxCost(IndexCache)
+	require.NoError(t, err)
+	require.Equal(t, int64(2<<20), cost)
+
+	require.NoError(t, db.UpdateCacheMaxCost(BlockCache, 2<<20))
+	cost, err = db.CacheMaxCost(BlockCache)
+	require.Equal(t, int64(2<<20), cost)
+	require.NoError(t, db.UpdateCacheMaxCost(IndexCache, 4<<20))
+	cost, err = db.CacheMaxCost(IndexCache)
```
Comment (jarifibrahim) [inline db_test.go:2213]: Missing error check.

File: db_test.go [hunk 3]
```diff
@@ -2184,3 +2184,32 @@ func TestWriteInemory(t *testing.T) {
...
+
+	cost, err := db.CacheMaxCost(BlockCache)
+	require.NoError(t, err)
+	require.Equal(t, int64(1<<20), cost)
+	cost, err = db.CacheMaxCost(IndexCache)
+	require.NoError(t, err)
+	require.Equal(t, int64(2<<20), cost)
+
+	require.NoError(t, db.UpdateCacheMaxCost(BlockCache, 2<<20))
+	cost, err = db.CacheMaxCost(BlockCache)
```
Comment (jarifibrahim) [inline db_test.go:2210]: missing error check.

File: go.sum [hunk 4]
```diff
@@ -15,6 +15,10 @@ github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/dgraph-io/ristretto v0.0.4-0.20200906165740-41ebdbffecfd h1:KoJOtZf+6wpQaDTuOWGuo61GxcPBIfhwRxRTaTWGCTc=
 github.com/dgraph-io/ristretto v0.0.4-0.20200906165740-41ebdbffecfd/go.mod h1:YylP9MpCYGVZQrly/j/diqcdUetCRRePeBB0c2VGXsA=
+github.com/dgraph-io/ristretto v0.0.4-0.20201001195931-875bfd715994 h1:7ePCMsbm/r1w3yCwUlp0BWJIFJYj//BYgza+RbNM8yc=
```
Comment (jarifibrahim) [inline go.sum:18]: You might need to run `go mod tidy`.

**Review by jarifibrahim — APPROVED** [review]

**Review by manishrjain — APPROVED** [review]
:lgtm: Let's bring it down to one API.

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1551#-:-MJ2R_2D6mxgyX_fpjBt:b-r42rnw)* status: all files reviewed, 5 unresolved discussions (waiting on @ashish-goswami, @jarifibrahim, and @martinmr)

___
*[db.go, line 1992 at r1](https://reviewable.io/reviews/dgraph-io/badger/1551#-MIaNJoYYlRpGZEuRHnu-r1-1992:-MJ2RO6p4HkIqEpSJJrg:bupumrl) ([raw file](https://github.com/dgraph-io/badger/blob/54d6a799172e527a45679ac1e7d7327b224a6711/db.go#L1992)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add a comment please since this is a public API.
</blockquote></details>

Don't need this API.
___
*[db.go, line 2051 at r2](https://reviewable.io/reviews/dgraph-io/badger/1551#-MJ2RJU56u69xYDK95M_:-MJ2RJU66Nh4RLKagzhM:b-qxevd6) ([raw file](https://github.com/dgraph-io/badger/blob/ed7ecaecb0ff1d9beaf2ce8bb5b8dc7679dee529/db.go#L2051)):*
> ```Go
> // The call will have an effect only if the DB was created with the cache. Otherwise it is
> // a no-op.
> func (db *DB) UpdateCacheMaxCost(cache CacheType, maxCost int64) error {
> ```

Call it CacheMaxCost. If maxCost < 0, don't update. just return the current value. Otherwise, update.

return would return the latest value.

**Review by martinmr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1551)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @ashish-goswami, @jarifibrahim, and @manishrjain)

___
*[db.go, line 1992 at r1](https://reviewable.io/reviews/dgraph-io/badger/1551#-MIaNJoYYlRpGZEuRHnu-r1-1992:-MJ45UQF7Myk_A2MhW8-:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/54d6a799172e527a45679ac1e7d7327b224a6711/db.go#L1992)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need this API.
</blockquote></details>

Done.
___
*[db.go, line 2051 at r2](https://reviewable.io/reviews/dgraph-io/badger/1551#-MJ2RJU56u69xYDK95M_:-MJ45V-M1jAsYOcb_4ag:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/ed7ecaecb0ff1d9beaf2ce8bb5b8dc7679dee529/db.go#L2051)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Call it CacheMaxCost. If maxCost < 0, don't update. just return the current value. Otherwise, update.

return would return the latest value.
</blockquote></details>

Done.
___
*[db_test.go, line 2210 at r1](https://reviewable.io/reviews/dgraph-io/badger/1551#-MIaNJoYYlRpGZEuRHnv-r1-2210:-MJ45X-A0SkMOl9ioI1Z:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/54d6a799172e527a45679ac1e7d7327b224a6711/db_test.go#L2210)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

missing error check.
</blockquote></details>

Done.
___
*[db_test.go, line 2213 at r1](https://reviewable.io/reviews/dgraph-io/badger/1551#-MIaNJoYYlRpGZEuRHnv-r1-2213:-MJ45XGx5UlOjoVALKQV:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/54d6a799172e527a45679ac1e7d7327b224a6711/db_test.go#L2213)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Missing error check.
</blockquote></details>

Done.
___
*[go.sum, line 18 at r1](https://reviewable.io/reviews/dgraph-io/badger/1551#-MIaNJoYYlRpGZEuRHnx-r1-18:-MJ45YDfCfpntimzxSNg:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/54d6a799172e527a45679ac1e7d7327b224a6711/go.sum#L18)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

You might need to run `go mod tidy`.
</blockquote></details>

Done.

**Pull request merged**